sound/soc/sof/ipc4-topology.c

```diff
@@ -1682,8 +1682,26 @@ static int sof_ipc4_prepare_process_module(struct snd_sof_widget *swidget,
 	}
 
 	control_data = sof_ipc4_get_widget_init_byte_kcontrol(swidget);
-	if (control_data)
+	if (control_data) {
+		if (fw_module->ext_library) {
+			struct sof_ipc4_base_module_cfg_ext *base_ext;
+			int size;
+
+			base_ext = (struct sof_ipc4_base_module_cfg_ext *)control_data->data->data;
+			size = sizeof(struct sof_ipc4_base_module_cfg_ext) +
+								base_ext->priv_param_len +
+								(base_ext->num_sink_pin_fmts +
+								base_ext->num_source_pin_fmts) *
+								sizeof(struct sof_ipc4_pin_format);
+			if (size != control_data->data->size) {
+				dev_err(sdev->dev, "Insufficient data for widget %s initialization\n",
+					swidget->widget->name);
+				return -EINVAL;
```

Comment:
how is this even possible? You call the exact same function during setup to allocate memory based on the size. How is it magically going to change here and suddenly become insufficient? So, to avoid this type of paranoia checks, please unify the calls to that function sof_ipc4_get_widget_init_byte_kcontrol() to just one place, so allocate and copy at the same time.

Reply:
the message is kind of confusing, I think I can make it `Invalid data for base module cofiguration extension`. the code here is to make sure the blob from topology fits in to base module config extension exactly.